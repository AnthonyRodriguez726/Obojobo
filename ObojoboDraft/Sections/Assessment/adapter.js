import ScoreActions from './post-assessment/score-actions'
// @TODO: Importing from the server code, we shouldn't do this:
import AssessmentRubric from '../../../server/assessment-rubric'

let Adapter = {
	construct(model, attrs) {
		if (attrs && attrs.content && attrs.content.scoreActions) {
			model.modelState.scoreActions = new ScoreActions(attrs.content.scoreActions)
		} else {
			model.modelState.scoreActions = new ScoreActions()
		}

		if (attrs && attrs.content && attrs.content.rubric) {
			model.modelState.rubric = new AssessmentRubric(attrs.content.rubric)
		} else {
			model.modelState.rubric = new AssessmentRubric()
		}

		model.setStateProp(
			'attempts',
			Infinity,
			p => (('' + p).toLowerCase() === 'unlimited' ? Infinity : parseInt(p, 10))
		)
		model.setStateProp('review', 'never', p => p.toLowerCase(), [
			'never',
			'always',
			'no-attempts-remaining'
		])
	},

	clone(model, clone) {
		clone.modelState.attempts = model.modelState.attempts
		clone.modelState.scoreActions = model.modelState.scoreActions.clone()
		clone.modelState.rubric = model.modelState.rubric.clone()
	},

	//@TODO - necessary?
	// clone.modelState.assessmentState =
	// 	inTest: model.modelState.assessmentState.inTest
	// 	currentScore: model.modelState.assessmentState.currentScore
	// 	scores: Object.assign [], model.modelState.assessmentState.scores

	toJSON(model, json) {
		json.content.attempts = model.modelState.attempts
		json.content.scoreActions = model.modelState.scoreActions.toObject()
		json.content.rubric = model.modelState.rubric.toObject()
	}
}

export default Adapter
