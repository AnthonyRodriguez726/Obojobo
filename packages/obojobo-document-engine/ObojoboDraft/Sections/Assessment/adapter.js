import ScoreActions from './score-actions'
// @TODO: Importing from the server code, we shouldn't do this:
import AssessmentRubric from '../../../server/assessment-rubric'

let Adapter = {
	construct(model, attrs) {
		// Default state.
		model.modelState.attempts = Infinity
		model.modelState.review = 'never'
		model.modelState.scoreActions = new ScoreActions()
		model.modelState.rubric = new AssessmentRubric(attrs.content.rubric)

		// Set state if XML has the attributes.
		if (attrs && attrs.content) {
			model.modelState.attempts =
				attrs.content.attempts === 'unlimited' ? Infinity : parseInt(attrs.content.attempts, 10)
			model.modelState.review = attrs.content.review || 'never'
			model.modelState.scoreActions = new ScoreActions(attrs.content.scoreActions || null)
		}
	},

	// model.modelState.assessmentState =
	// 	inTest: false
	// 	scores: []
	// 	currentScore: 0

	clone(model, clone) {
		clone.modelState.attempts = model.modelState.attempts
		clone.modelState.hideNav = model.modelState.hideNav
		clone.modelState.scoreActions = model.modelState.scoreActions.clone()
		clone.modelState.rubric = model.modelState.rubric.clone()
	},

	//@TODO - necessary?
	// clone.modelState.assessmentState =
	// 	inTest: model.modelState.assessmentState.inTest
	// 	currentScore: model.modelState.assessmentState.currentScore
	// 	scores: Object.assign [], model.modelState.assessmentState.scores

	toJSON(model, json) {
		json.content.attempts = model.modelState.attempts
		json.content.hideNav = model.modelState.hideNav
		json.content.scoreActions = model.modelState.scoreActions.toObject()
		json.content.rubric = model.modelState.rubric.toObject()
	}
}

export default Adapter
