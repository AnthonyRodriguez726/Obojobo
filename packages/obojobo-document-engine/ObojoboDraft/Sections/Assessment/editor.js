import React from 'react'
import { Block } from 'slate'
import { CHILD_REQUIRED, CHILD_TYPE_INVALID } from 'slate-schema-violations'

const ASSESSMENT_NODE = 'ObojoboDraft.Sections.Assessment'
const SETTINGS_NODE = 'ObojoboDraft.Sections.Assessment.Settings'
const RUBRIC_NODE = 'ObojoboDraft.Sections.Assessment.Rubric'
const ACTIONS_NODE = 'ObojoboDraft.Sections.Assessment.ScoreActions'
const QUESTION_BANK_NODE = 'ObojoboDraft.Chunks.QuestionBank'
const PAGE_NODE = 'ObojoboDraft.Pages.Page'

import Page from '../../Pages/Page/editor'
import QuestionBank from '../../Chunks/QuestionBank/editor'
import ScoreActions from './post-assessment/editor'
import Rubric from './components/rubric/editor'
import ParameterNode from '../../../src/scripts/oboeditor/components/parameter-node'

const Settings = props => {
	return (
		<div className={'settings pad'}>
			<span contentEditable={false} className={'label'}>
				{'Assessment Settings'}
			</span>
			<div>{props.children}</div>
		</div>
	)
}

class Node extends React.Component {
	constructor(props) {
		super(props)
		this.state = this.props.node.data.get('content')
	}

	addRubric() {
		const editor = this.props.editor
		const change = editor.value.change()

		const newRubric = Block.create({
			type: RUBRIC_NODE,
			data: { content: { type: 'pass-fail' } }
		})
		change.insertNodeByKey(this.props.node.key, this.props.node.nodes.size, newRubric)

		editor.onChange(change)
	}

	render() {
		const hasRubric = this.props.node.nodes.size === 5
		return (
			<div className={'assessment'}>
				{this.props.children}
				{!hasRubric ? (
					<button className={'add-rubric'} onClick={() => this.addRubric()}>
						{'Add Rubric'}
					</button>
				) : null}
			</div>
		)
	}
}

const slateToObo = node => {
	const json = {}
	json.id = node.key
	json.type = node.type
	json.content = node.data.get('content')
	json.children = []

	// Remove rubric if it has been deleted
	delete json.content.rubric

	node.nodes.forEach(child => {
		switch (child.type) {
			case PAGE_NODE:
				json.children.push(Page.helpers.slateToObo(child))
				break
			case QUESTION_BANK_NODE:
				json.children.push(QuestionBank.helpers.slateToObo(child))
				break
			case ACTIONS_NODE:
				json.content.scoreActions = ScoreActions.helpers.slateToObo(child)
				break
			case RUBRIC_NODE:
				json.content.rubric = Rubric.helpers.slateToObo(child)
				break
			case SETTINGS_NODE:
				json.content.attempts = child.nodes.get(0).text
				json.content.review = child.nodes.get(1).data.get('current')
		}
	})

	return json
}

const oboToSlate = node => {
	const json = {}
	json.object = 'block'
	json.key = node.id
	json.type = ASSESSMENT_NODE
	json.data = { content: node.get('content') }
	json.nodes = []

	const settings = {
		object: 'block',
		type: SETTINGS_NODE,
		nodes: []
	}

	settings.nodes.push(
		ParameterNode.helpers.oboToSlate({
			name: 'attempts',
			value: json.data.content.attempts + '',
			display: 'Attempts'
		})
	)

	settings.nodes.push(
		ParameterNode.helpers.oboToSlate({
			name: 'review',
			value: json.data.content.review,
			display: 'Review',
			options: ['always', 'never', 'no-attempts-remaining']
		})
	)

	json.nodes.push(settings)

	node.attributes.children.forEach(child => {
		if (child.type === PAGE_NODE) {
			json.nodes.push(Page.helpers.oboToSlate(child))
		} else {
			json.nodes.push(QuestionBank.helpers.oboToSlate(child))
		}
	})

	json.nodes.push(ScoreActions.helpers.oboToSlate(json.data.content.scoreActions))
	if (json.data.content.rubric) json.nodes.push(Rubric.helpers.oboToSlate(json.data.content.rubric))

	return json
}

const plugins = {
	renderNode(props) {
		switch (props.node.type) {
			case ASSESSMENT_NODE:
				return <Node {...props} {...props.attributes} />
			case SETTINGS_NODE:
				return <Settings {...props} {...props.attributes} />
		}
	},
	validateNode(node) {
		if (node.object !== 'block') return
		if (node.type !== SETTINGS_NODE) return
		if (node.nodes.first().object === 'text') return
		if (node.nodes.size !== 1) return
		if (node.nodes.first().data.get('name') === 'attempts') return

		const block = Block.create({
			type: 'Parameter',
			data: {
				name: 'attempts',
				display: 'Attempts'
			}
		})

		return change => change.insertNodeByKey(node.key, 0, block)
	},
	schema: {
		blocks: {
			'ObojoboDraft.Sections.Assessment': {
				nodes: [
					{ match: [{ type: SETTINGS_NODE }], min: 1, max: 1 },
					{ match: [{ type: PAGE_NODE }], min: 1, max: 1 },
					{ match: [{ type: QUESTION_BANK_NODE }], min: 1, max: 1 },
					{ match: [{ type: ACTIONS_NODE }], min: 1, max: 1 },
					{ match: [{ type: RUBRIC_NODE }], max: 1 }
				],
				normalize: (change, error) => {
					const { node, child, index } = error
					switch (error.code) {
						case CHILD_REQUIRED: {
							let block
							switch (index) {
								case 0:
									block = Block.create({
										type: SETTINGS_NODE
									})
									break
								case 1:
									block = Block.create({
										type: PAGE_NODE
									})
									break
								case 2:
									block = Block.create({
										type: QUESTION_BANK_NODE,
										data: { content: { choose: 1, select: 'sequential' } }
									})
									break
								case 3:
									block = Block.create({
										type: ACTIONS_NODE
									})
									break
							}
							return change.insertNodeByKey(node.key, index, block)
						}
						case CHILD_TYPE_INVALID: {
							switch (index) {
								case 0:
									return change.wrapBlockByKey(child.key, {
										type: SETTINGS_NODE
									})
								case 1:
									return change.wrapBlockByKey(child.key, {
										type: PAGE_NODE
									})
								case 2:
									return change.wrapBlockByKey(child.key, {
										type: QUESTION_BANK_NODE,
										data: { content: { choose: 1, select: 'sequential' } }
									})
								case 3:
									return change.wrapBlockByKey(child.key, {
										type: ACTIONS_NODE
									})
							}
						}
					}
				}
			},
			'ObojoboDraft.Sections.Assessment.Settings': {
				nodes: [{ match: [{ type: 'Parameter' }], min: 2, max: 2 }],
				normalize: (change, error) => {
					const { node, child, index } = error
					switch (error.code) {
						case CHILD_REQUIRED: {
							if (index === 0) {
								const block = Block.create(
									ParameterNode.helpers.oboToSlate({
										name: 'attempts',
										value: 'unlimited',
										display: 'Attempts'
									})
								)
								return change.insertNodeByKey(node.key, index, block)
							}
							const block = Block.create(
								ParameterNode.helpers.oboToSlate({
									name: 'review',
									value: 'never',
									display: 'Review',
									options: ['always', 'never', 'no-attempts-remaining']
								})
							)
							return change.insertNodeByKey(node.key, index, block)
						}
						case CHILD_TYPE_INVALID: {
							return change.withoutNormalization(c => {
								c.removeNodeByKey(child.key)
								if (index === 0) {
									const block = Block.create(
										ParameterNode.helpers.oboToSlate({
											name: 'attempts',
											value: 'unlimited',
											display: 'Attempts'
										})
									)
									return c.insertNodeByKey(node.key, index, block)
								}
								const block = Block.create(
									ParameterNode.helpers.oboToSlate({
										name: 'review',
										value: 'never',
										display: 'Review',
										options: ['always', 'never', 'no-attempts-remaining']
									})
								)
								return c.insertNodeByKey(node.key, index, block)
							})
						}
					}
				}
			}
		}
	}
}

const Assessment = {
	components: {
		Node,
		Settings
	},
	helpers: {
		slateToObo,
		oboToSlate
	},
	plugins
}

export default Assessment
