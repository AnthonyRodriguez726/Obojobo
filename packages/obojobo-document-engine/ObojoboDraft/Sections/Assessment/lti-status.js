import './lti-status.scss'

import Common from 'Common'
import Viewer from 'Viewer'

let { Button } = Common.components
let LTINetworkStates = Viewer.stores.assessmentStore.LTINetworkStates

export default props => {
	if (!props.ltiState.state) return null

	switch (props.ltiState.state.gradebookStatus) {
		case 'ok_no_outcome_service':
		case 'ok_gradebook_matches_assessment_score':
		case 'ok_null_score_not_sent':
			return null
	}

	let systemLabel = props.externalSystemLabel
	return (
		<div className="obojobo-draft--sections--assessment--lti-status">
			<h2>{`There was a problem sending your score to ${systemLabel}.`}</h2>
			<p>
				{`Don’t worry - your score is safely recorded here. We just weren’t able to send it to ${systemLabel}. Click the button below to resend your score:`}
			</p>
			{props.ltiState.errorCount === 0 ||
			props.ltiState.networkState !== LTINetworkStates.IDLE ? null : (
				<p>
					<strong>Sorry - That didn't work.</strong>
					{` Most likely the connection to ${systemLabel} has expired and just needs to be refreshed. Please close this tab or window, reopen this module from ${systemLabel}, return to this page and then resend your score.`}
				</p>
			)}
			{(() => {
				switch (props.ltiState.networkState) {
					case LTINetworkStates.AWAITING_SEND_ASSESSMENT_SCORE_RESPONSE:
						return <Button disabled>Resending Score...</Button>
						break

					case LTINetworkStates.IDLE:
					default:
						return (
							<Button dangerous onClick={props.onClickResendScore}>
								{props.ltiState.errorCount === 0 ? 'Resend score' : 'Try again anyway'}
							</Button>
						)
						break
				}
			})()}
		</div>
	)
}
