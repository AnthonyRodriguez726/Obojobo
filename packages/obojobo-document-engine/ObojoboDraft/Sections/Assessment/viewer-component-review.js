import Viewer from 'Viewer'
import Common from 'Common'

const { AssessmentUtil } = Viewer.util
const { NavUtil } = Viewer.util
const { OboModel } = Common.models

const assessmentReviewView = assessment => {
	let attemptReviewComponents = {}

	let attempts = AssessmentUtil.getAllAttempts(
		assessment.props.moduleData.assessmentState,
		assessment.props.model
	)

	let attemptReviewComponent = (attempt, assessment) => {
		return (
			<div className="review">
				<h1>{`Attempt ${attempt.attemptNumber}`}</h1>
				<h2>{`Score: ${attempt.attemptScore}`}</h2>
				{attempt.questionScores.map(scoreObj => {
					const questionModel = OboModel.models[scoreObj.id]
					const QuestionComponent = questionModel.getComponentClass()
					return (
						<QuestionComponent
							model={questionModel}
							moduleData={assessment.props.moduleData}
							mode={'review'}
						/>
					)
				})}
			</div>
		)
	}

	attempts.forEach(attempt => {
		attemptReviewComponents[`assessmentReview:${attempt.attemptId}`] = attemptReviewComponent(
			attempt,
			assessment
		)
	})

	let context = assessment.props.moduleData.navState.context
	if (context.split(':')[0] !== 'assessmentReview')
		// show most recent attempt
		NavUtil.setContext(`assessmentReview:${attempts[attempts.length - 1].attemptId}`)

	let attemptButtons = attempts.map((attempt, index) => {
		return (
			<button onClick={() => NavUtil.setContext(`assessmentReview:${attempt.attemptId}`)}>
				Attempt #{attempt.attemptNumber}
			</button>
		)
	})

	return (
		<div className="score unlock">
			{attemptButtons}
			{attemptReviewComponents[context]}
			{attemptButtons}
		</div>
	)
}

export default assessmentReviewView
