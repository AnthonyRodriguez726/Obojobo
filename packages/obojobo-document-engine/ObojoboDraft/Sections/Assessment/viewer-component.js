import './viewer-component.scss'

import React from 'react'

import Common from 'Common'
import Viewer from 'Viewer'

const { OboComponent } = Viewer.components
const { Dispatcher } = Common.flux
const { ModalUtil } = Common.util

const { AssessmentUtil } = Viewer.util
const { NavUtil, FocusUtil } = Viewer.util

import AttemptIncompleteDialog from './components/attempt-incomplete-dialog'
import { FOCUS_ON_ASSESSMENT_CONTENT } from './assessment-event-constants'

import PreTest from './components/pre-test'
import Test from './components/test'
import PostTest from './components/post-test'

class Assessment extends React.Component {
	static focusOnContent() {
		Dispatcher.trigger(FOCUS_ON_ASSESSMENT_CONTENT)
	}

	static getCurrentStep(model, assessmentState) {
		const assessment = AssessmentUtil.getAssessmentForModel(assessmentState, model)

		if (assessment === null) {
			return 'pre-test'
		}

		if (assessment.current !== null) {
			return 'test'
		}

		if (assessment.attempts.length > 0) {
			return 'post-test'
		}

		return 'pre-test'
	}

	constructor() {
		super()
		this.state = {
			isFetching: false,
			step: null
		}

		this.onEndAttempt = this.onEndAttempt.bind(this)
		this.onAttemptEnded = this.onAttemptEnded.bind(this)
		this.endAttempt = this.endAttempt.bind(this)
		this.onClickSubmit = this.onClickSubmit.bind(this)

		Dispatcher.on('assessment:endAttempt', this.onEndAttempt)
		Dispatcher.on('assessment:attemptEnded', this.onAttemptEnded)
	}

	getCurrentStep() {
		return this.constructor.getCurrentStep(this.props.model, this.props.moduleData.assessmentState)
	}

	componentWillReceiveProps() {
		// If we were on a step (meaning we haven't left the assessment section)
		// and the step has changed we need to scroll to the top and focus
		// on the content
		const curStep = this.getCurrentStep()

		if (this.state.step !== null && curStep !== this.state.step) {
			this.needsScrollToTopAndFocus = true
		}

		this.setState({
			step: curStep
		})
	}

	componentWillUnmount() {
		NavUtil.setContext('practice')
		Dispatcher.off('assessment:endAttempt', this.onEndAttempt)
		Dispatcher.off('assessment:attemptEnded', this.onAttemptEnded)
	}

	componentDidUpdate() {
		if (this.needsScrollToTopAndFocus) {
			delete this.needsScrollToTopAndFocus
			Dispatcher.trigger('viewer:scrollToTop')

			FocusUtil.focusOnNavTargetContent()
		}
	}

	onEndAttempt() {
		this.setState({ isFetching: true })
	}

	onAttemptEnded() {
		this.setState({ isFetching: false })
	}

	isAttemptComplete() {
		return AssessmentUtil.isCurrentAttemptComplete(
			this.props.moduleData.assessmentState,
			this.props.moduleData.questionState,
			this.props.model,
			this.props.moduleData.navState.context
		)
	}

	isAssessmentComplete() {
		return !AssessmentUtil.hasAttemptsRemaining(
			this.props.moduleData.assessmentState,
			this.props.model
		)
	}

	onClickSubmit() {
		// disable multiple clicks
		if (this.state.isFetching) return

		if (!this.isAttemptComplete()) {
			ModalUtil.show(<AttemptIncompleteDialog onSubmit={this.endAttempt} />)
			return
		}
		return this.endAttempt()
	}

	endAttempt() {
		return AssessmentUtil.endAttempt({
			model: this.props.model,
			context: this.props.moduleData.navState.context,
			visitId: this.props.moduleData.navState.visitId
		})
	}

	exitAssessment() {
		const scoreAction = this.getScoreAction()

		switch (scoreAction.action.value) {
			case '_next':
				return NavUtil.goNext()

			case '_prev':
				return NavUtil.goPrev()

			default:
				return NavUtil.goto(scoreAction.action.value)
		}
	}

	getScoreAction() {
		const assessmentScore = AssessmentUtil.getAssessmentScoreForModel(
			this.props.moduleData.assessmentState,
			this.props.model
		)
		const scoreAction = this.props.model.modelState.scoreActions.getActionForScore(assessmentScore)

		if (scoreAction) {
			return scoreAction
		}

		return {
			from: 0,
			to: 100,
			message: '',
			action: {
				type: 'unlock',
				value: '_next'
			}
		}
	}

	render() {
		const childEl = (() => {
			switch (this.getCurrentStep()) {
				case 'pre-test':
					return (
						<PreTest model={this.props.model.children.at(0)} moduleData={this.props.moduleData} />
					)

				case 'test':
					return (
						<Test
							model={this.props.model.children.at(1)}
							moduleData={this.props.moduleData}
							onClickSubmit={this.onClickSubmit}
							isAttemptComplete={this.isAttemptComplete()}
							isFetching={this.state.isFetching}
						/>
					)

				case 'post-test':
					return (
						<PostTest
							ref="child"
							model={this.props.model}
							moduleData={this.props.moduleData}
							scoreAction={this.getScoreAction()}
						/>
					)
			}
		})()

		return (
			<OboComponent
				model={this.props.model}
				moduleData={this.props.moduleData}
				className="obojobo-draft--sections--assessment"
			>
				{childEl}
			</OboComponent>
		)
	}
}

export default Assessment
