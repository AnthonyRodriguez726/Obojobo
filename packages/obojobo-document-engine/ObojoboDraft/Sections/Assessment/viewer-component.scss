@import '~styles/includes';

.obojobo-draft--sections--assessment
{
	.untested,
	.score
	{
		// text-align: center;
		padding-top: 1em;

		> h1
		{
			margin-bottom: 0.2em;
			text-align: center;
		}

		> h2
		{
			margin-bottom: 0;
			text-align: center;
			font-size: 1em;
		}

		p
		{
			margin-bottom: 3em;
		}
	}

	.review
	{
		text-align: left;
		width: 70%;
		margin: 0 auto;
		border: 1px solid $color-shadow;
		padding: 3em;
		margin-top: 3em;
		background: #fdfdfd;
		border-radius: .2em;

		.number-correct
		{
			font-family: $font-default;
			margin-top: 0;
			margin-bottom: 2em;
		}

		.is-correct > p,
		.is-not-correct > p
		{
			margin-top: 0;
			margin-bottom: 0;
			font-family: $font-default;
			font-weight: bold;
		}

		.obojobo-draft--chunks--question
		{
			font-size: 0.8em;
		}

		.is-not-correct .obojobo-draft--chunks--question .flipper
		{
			position: relative;
			border: 1px solid $color-incorrect;
			background: transparentize($color-incorrect, 0.9);

			.responses-hidden
			{
				color: $color-incorrect;
			}
		}

		.is-correct .obojobo-draft--chunks--question .flipper
		{
			position: relative;
			border: 1px solid $color-correct;
			background: transparentize($color-correct, 0.9);

			.responses-hidden
			{
				color: $color-correct;
			}
		}
	}

	.submit-button
	{
		display: block;
		margin: 0 auto;
		// transform: scale(1.25, 1.25);
		text-align: center;
		width: $dimension-column-width;

		button {
			min-width: 9em;
		}
	}
}