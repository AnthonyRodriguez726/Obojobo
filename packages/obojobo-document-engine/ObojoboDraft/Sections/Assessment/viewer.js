import adapter from './adapter'
import Common from 'Common'
import Viewer from 'Viewer'
import ViewerComponent from './viewer-component'

const { AssessmentUtil } = Viewer.util

Common.Registry.registerModel('ObojoboDraft.Sections.Assessment', {
	adapter: adapter,
	componentClass: ViewerComponent,
	type: 'section',
	getNavItem(model) {
		const title = model.title || 'Assessment'

		return {
			type: 'link',
			label: title,
			path: [title.toLowerCase().replace(/ /g, '-')],
			showChildren: false,
			showChildrenOnNavigation: false
		}
	},
	variables: {
		'assessment:attemptsRemaining'(textModel, viewerProps) {
			const assessmentModel = textModel.getParentOfType('ObojoboDraft.Sections.Assessment')
			if (assessmentModel.modelState.attempts === Infinity) {
				return 'unlimited'
			}

			return AssessmentUtil.getAttemptsRemaining(viewerProps.assessmentState, assessmentModel)
		},
		'assessment:attemptsAmount'(textModel) {
			const assessmentModel = textModel.getParentOfType('ObojoboDraft.Sections.Assessment')
			if (assessmentModel.modelState.attempts === Infinity) {
				return 'unlimited'
			}

			return assessmentModel.modelState.attempts
		}
	}
})
