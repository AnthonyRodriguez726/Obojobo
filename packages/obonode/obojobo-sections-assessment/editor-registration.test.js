jest.mock('./editor-component', () => global.mockReactComponent(this, 'Assessment'))
jest.mock('./components/settings/editor-component', () =>
	global.mockReactComponent(this, 'Settings')
)
jest.mock('./schema', () => ({ mock: 'schema' }))
jest.mock('./converter', () => ({ mock: 'converter' }))
jest.mock('slate-react')

import SlateReact from 'slate-react'
import Assessment from './editor-registration'

const ASSESSMENT_NODE = 'ObojoboDraft.Sections.Assessment'
const SETTINGS_NODE = 'ObojoboDraft.Sections.Assessment.Settings'

describe('Assessment editor', () => {
	test.skip('plugins.onPaste pastes anything other than an Assessment', () => {
		SlateReact.getEventTransfer.mockReturnValueOnce({
			type: 'fragment',
			fragment: {
				nodes: {
					get: () => ({ type: 'mockNode' })
				}
			}
		})
		const next = jest.fn()

		Assessment.plugins.onPaste(null, null, next)

		expect(next).toHaveBeenCalled()
	})

	test.skip('plugins.onPaste pastes an Assessment', () => {
		const assessmentMock = {
			type: ASSESSMENT_NODE,
			toJSON: jest.fn().mockReturnValueOnce('mock assessment')
		}

		SlateReact.getEventTransfer.mockReturnValueOnce({
			type: 'fragment',
			fragment: {
				nodes: {
					get: () => assessmentMock
				}
			}
		})

		const editor = {
			insertFragment: jest.fn()
		}

		Assessment.plugins.onPaste(null, editor, jest.fn())

		expect(editor.insertFragment).toHaveBeenCalled()
	})

	test('plugins.renderNode renders the Assessment when passed', () => {
		const props = {
			attributes: { dummy: 'dummyData' },
			node: {
				type: ASSESSMENT_NODE,
				data: {
					get: () => {
						return {}
					}
				}
			}
		}

		expect(Assessment.plugins.renderNode(props, null, jest.fn())).toMatchSnapshot()
	})

	test('plugins.renderNode renders Settings when passed', () => {
		const props = {
			attributes: { dummy: 'dummyData' },
			node: {
				type: SETTINGS_NODE,
				data: {
					get: () => {
						return {}
					}
				}
			}
		}

		expect(Assessment.plugins.renderNode(props, null, jest.fn())).toMatchSnapshot()
	})

	test('plugins.renderNode calls next', () => {
		const props = {
			attributes: { dummy: 'dummyData' },
			node: {
				type: 'mockNode',
				data: {
					get: () => {
						return {}
					}
				}
			}
		}

		const next = jest.fn()

		expect(Assessment.plugins.renderNode(props, null, next)).toMatchSnapshot()
		expect(next).toHaveBeenCalled()
	})

	test('getNavItem returns expected object', () => {
		const model = {
			parent: {
				children: {
					models: [{ get: () => true }]
				}
			},
			title: 'Test Title'
		}

		expect(Assessment.getNavItem(model)).toEqual({
			type: 'link',
			label: 'Test Title',
			path: ['test-title'],
			showChildren: false,
			showChildrenOnNavigation: false
		})

		model.title = null
		expect(Assessment.getNavItem(model)).toEqual({
			type: 'link',
			label: 'Assessment',
			path: ['assessment'],
			showChildren: false,
			showChildrenOnNavigation: false
		})
	})
})
