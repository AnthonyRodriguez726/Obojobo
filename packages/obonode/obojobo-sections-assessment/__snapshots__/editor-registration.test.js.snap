// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Assessment editor plugins.renderNode calls next 1`] = `undefined`;

exports[`Assessment editor plugins.renderNode renders Settings when passed 1`] = `
<Settings
  attributes={
    Object {
      "dummy": "dummyData",
    }
  }
  dummy="dummyData"
  node={
    Object {
      "data": Object {
        "get": [Function],
      },
      "type": "ObojoboDraft.Sections.Assessment.Settings",
    }
  }
/>
`;

exports[`Assessment editor plugins.renderNode renders the Assessment when passed 1`] = `
<Assessment
  attributes={
    Object {
      "dummy": "dummyData",
    }
  }
  dummy="dummyData"
  node={
    Object {
      "data": Object {
        "get": [Function],
      },
      "type": "ObojoboDraft.Sections.Assessment",
    }
  }
/>
`;
